SFDP: Add support for multiple configurations and sector maps

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes #12574

This PR adds to following changes, to support CYW9P62S1_43012EVB_01's QSPI flash (S25FS512S):

* A Sector Map Parameter Table contains a sequence of the following descriptors:
  * (Optional) configuration detection command descriptors, one for each command to run to determine the current configuration. This exists only if the flash layout is configurable.
  * Sector map descriptors, one map for each possible configuration. On a flash device with a non-configurable layout, there is only one such descriptor/map.

  Previously we only supported the non-configurable case with a single descriptor. This PR adds support for multiple configurations and provides a positive unit test case.

* To be able to run configuration detection commands, a callback provided by each block device and used by the SFDP class needs to take additional parameters: address size, instruction and dummy cycles. Their values come from sector map descriptors in the SFDP table. This PR changes the callback interface, and updates callbacks implemented by SPI, QSPI and OSPI block devices to the extended interface.
* A quirk is provided for Cypress S25FS512S flash device's inconsistent register value: This device's SFDP table suggests checking CR3NV[1], which needs to be 1 in order for the config detection algorithm to work. But the actual hardware has value 0 instead. This has been discussed before on the Internet (e.g. on [a Linux mailing list](https://linux-mtd.infradead.narkive.com/ylHK6CyT/spi-nor-fs512s-incorrect-cr3nv-1-value)) - it's not possible to change a value in hardware that has been in production for years.
* Two of the three configurations of S25FS512S have eight 4KB sectors (32KB in total) overlaying a 256KB sector. This type of partial overlay is not current supported by Mbed OS's BlockDevice and SFDP parser. A quirk is provided to ignore the 4KB sectors and only use the 256KB sectors.
* Fixes of a few preexisting issues we found.

More details in commit messages.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

To check.

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
SFDP: Add more parameters to the reader callback

The SFDP functions parse SFDP data which is fetched by a callback
called `sfdp_reader` provided by {SPIF,QSPIF,OSPIF}BlockDevice.
Currently, this callback interface only takes a read address and an RX
buffer to store output data. This has been enough, because other SPI
parameters are always the same when fetching the SFDP table only -
they are just hardcoded in each reader.

But in the future we will add support for flash devices with multiple
configurations (in a subsequent commit), and to detect which
configuration is enabled, we will need to send detection commands
which require device-dependent SPI parameters:
* address size
* instruction
* dummy cycles

This commit
* turns the above SPI parameters from predefined/hardcoded values
into parameters of the callback
* lets the SFDP functions pass the above parameters to the callback
(Note: To read the SFDP table itself, those values are constants
defined by the standard, not tied to any particular device, so they
can be known to the SFDP functions)
* updates the callbacks implemented by {SPIF,QSPIF,OSPIF}BlockDevice
* updates the mock callback for unit tests and expectations

LDong-Arm committed:
SFDP: unit tests: Move unshared test data into case

Some test data in test_sfdp.cpp is used by one test case only, so
we turn it into a local variable.

LDong-Arm committed:
SFDP: Add support for multiple configurations and sector maps

A Sector Map Parameter Table contains a sequence of the following
descriptors:
* (Optional) configuration detection command descriptors, one for
each command to run to determine the current configuration. This
exists only if the flash layout is configurable.
* Sector map descriptors, one for each possible configuration. On
a flash device with a non-configurable layout, there is only one
such descriptor.

Previously we only supported the non-configurable case with a single
descriptor. This commit adds support for multiple configurations.

LDong-Arm committed:
SFDP: unit tests: Fix Configuration ID in fake test data

The second byte of the sector map descriptor is the configuration ID.
On a device with non-configurable layout, the only available map
descriptor's configuration ID must be 0x00 as required by the
JESD216D standard. This value is important, because we will check
each descriptor's configuration ID when we support multiple
configurations.

Note: The test data is fake - when we modified real data of a
configurable device to become non-configurable for test purpose, we
forgot to change this field.

LDong-Arm committed:
QSPIFBlockDevice: Add quirk to ignore overlaying sectors on S25FS512S

The entire flash chip S25FS512S consists of uniform 256KB sectors.
Additionally, it has three configurations:
* 0x01: Eight 4KB sectors (32KB in total) overlaying the start of
the first 256KB sector
* 0x03: Eight 4KB sectors (32KB in total) overlaying the end of the
last 256KB sector
* 0x05: No overlaying sectors

The active configuration is determined from bit fields of two
registers, CR1NV and CR3NV.

Mbed OS does not currently support partially overlaying sectors,
meaning that with eight 4KB sectors overlay a 256KB sectors, the
remaining 224KB (== 256KB - 8 * 4KB) of the big sector can't be
correctly handled. Supporting such scenario would involve a large
amount of rewriting in Mbed OS's BlockDevice, SFDP and their tests,
and may increase the code size.

So, this commit applies a quirk to always report configuration 0x05
(no overlaying sectors). Even if 0x01 or 0x03 is the real configuration,
they are compatible with the 0x05 configuration because 256KB sectors
exist in all cases.

Note: This quirk does *not* change actual configurations on hardware,
because registers CR1NV and CR3NV are one-time configurable (OTP) -
each bit field has a factory value of 0 and can be changed to 1 by the
user but not back to 0. So QSPIFBlockDevice avoids changing them.

Patater commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -1099,6 +1102,15 @@ int QSPIFBlockDevice::_handle_vendor_quirks()
+                // expected to be 1 but its actual value is 0 on manufactured hardware.
+                // In order for configuration detection to work, this chip needs to be
+                // specially handled.
+                _CR3NV_quirk = false;
```
This looks like we have detected a problematic S25FS512S device. Should this be `_CR3NV_quirk = true;`? Where do we set `_CR3NV_quirk` to `true` if anywhere?

LDong-Arm replied:
Good catch, it should be true here.

LDong-Arm replied:
Done

Patater commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -1469,6 +1481,14 @@ int QSPIFBlockDevice::_qspi_send_read_sfdp_command(mbed::bd_addr_t addr, mbed::s
+    if (_CR3NV_quirk && (addr == register_CR3NV)) {
+        // If we reach here, rx_buffer is guaranteed to be non-null
+        // because it's been checked by _qspi.read() above.
+        static_cast<uint8_t*>(rx_buffer)[0] |= 0x01;
```
We should add a unit test to catch this when our mocked QSPI device has a certain device ID and when it doesn't.

LDong-Arm replied:
We currently don't have unit tests for QSPIFBlockDevice. I'll create a test cpp file and add a test for this quirk (i.e. what we change in this PR).

LDong-Arm replied:
Unit test now added for the quirk. I eventually did it differently, because what we discussed didn't quite work - see commit message and inline comments for details.

LDong-Arm committed:
CYW9P62S1_43012EVB_01: Enable QSPI and QSPIF

QSPIF was disabled on CYW9P62S1_43012EVB_01 because its S25FS512S
flash chip has multiple configurations but our SFDP parser did not
support this scenario. Now having added support for this, we can
enable QSPIF (the component label for QSPIFBlockDevice) and QSPI
(the label for Quad-SPI communication).

rwalton-arm commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -1405,23 +1412,53 @@ qspi_status_t QSPIFBlockDevice::_qspi_send_general_command(qspi_inst_t instructi
+        case SFDP_CMD_ADDR_SIZE_VARIABLE: // use current setting
+            break;
+        case SFDP_CMD_ADDR_NONE: // no address in command
+            address = (int) QSPI_NO_ADDRESS_COMMAND;
```
Is it better to use `static_cast` here? C style casting will prevent the compiler from catching some potential errors.

LDong-Arm replied:
Good point, I prefer `static_cast` too and will switch to it.

LDong-Arm replied:
Done

rwalton-arm commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -1405,23 +1412,53 @@ qspi_status_t QSPIFBlockDevice::_qspi_send_general_command(qspi_inst_t instructi
+    // Don't check the read status until after we've configured the format back to 1-1-1,
+    // to avoid leaving the interface in an incorrect state if the read fails.
+    size_t rx_len = rx_length;
+    status = _qspi.read(inst, -1, (unsigned int) address, (char *) rx_buffer, &rx_len);
```
Preexisting, but I'm wondering if we should also use "c++ style" named casts here?

LDong-Arm replied:
I'll fix it. Also, `_qspi.read()` takes `int` as argument 2 so it should not be casted to `unsigned int`.

LDong-Arm replied:
Done

LDong-Arm committed:
QSPIFBlockDevice: Use fully-qualified include path

In Mbed OS, each library has an `include/<library>/` subdirectory
containing headers. The recommended way to include a header is
`#include "<library>/<header>.h"` to avoid potential conflicts with
any external modules that have same names of headers.

This is not enforced yet, and both include/ and include/<component>/
are in a library's include paths, to avoid breaking preexisting
Mbed projects that don't follow the recommendation. But code within
Mbed OS should follow it at least.

LDong-Arm committed:
QSPIFBlockDevice: Add unit test for the S25FS512S quirk

Add a test case to verify that the quirk on `CR1NV` and `CR3NV`
registers get applied by `QSPIFBlockDevice` when the flash device
is S25FS512S.

`QSPIFBlockDevice` depends on the SFDP functions and the `QSPI` class,
but we can't use gMock on them because:
* SFDP functions are global functions, whereas gMock only supports
mocking class member functions.
* A mocked class is injected into the test subject, but passing a
preexisting `QSPI` instance into `QSPIFBlockDevice` is not possible
(unless we resort to pointers). For details, see comments in
test_QSPIFBlockDevice.cpp.

So fakes of the `QSPI` class and any SFDP functions involved are
defined in test_QSPIFBlockDevice.cpp.

LDong-Arm committed:
QSPI: Move destructor into source file

This allows the entire QSPI class to be mocked/faked for unit testing
purpose, without dependencies from the real implementation such as
`qspi_free()` from the HAL.

jeromecoutant reviewed (changes requested):
Needs update for OSPI as well

LDong-Arm committed:
drivers: Add missing OSPI.cpp to CMakeLists.txt

Patater commented on storage/blockdevice/COMPONENT_QSPIF/UNITTESTS/test_QSPIFBlockDevice.cpp:
```diff
@@ -0,0 +1,262 @@
+ * We can't use gMock for this because
+ * - The entire mbed::QSPI is non-virtual. Mocks are derived classes, and
+ * when a derived class instance is used as an instance of its base class,
+ * non-virtual members of the base class get usedx.
```
s/usedx/used

LDong-Arm replied:
Done

Patater commented on storage/blockdevice/COMPONENT_QSPIF/UNITTESTS/test_QSPIFBlockDevice.cpp:
```diff
@@ -0,0 +1,262 @@
+ * instead of a reference, in order to support initializing from pins directly.
+ * Also, mbed::QSPI is declared as NonCopyable whose copy-constructor is deleted,
+ * so we can't copy a preexisting mbed::QSPI instance into QSPIFBlockDevice.
+ * There's no way to dependency inject a mock instance.
```
Could we change QSPIFBlockDevice's member to be a reference instead of a copy? This is an internal implementation detail and shouldn't affect use of the public API.

LDong-Arm replied:
The existing constructor is
```
QSPIFBlockDevice(PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName csel,
                                   int clock_mode,
                                   int freq);
```
so to support this constructor, the member `QSPI _qspi` needs to be a "full" one (instead of a reference) in order to construct a QSPI object from pins. If we change this member to a reference `QSPI &_qspi`, then QSPIFBlockDevice won't reserve memory for a full QSPI object, and the workaround I can think of would be using heap to construct a QSPI object from pins in the QSPIFBlockDevice constructor that takes pin numbers.

Patater replied:
A pointer allow us to inject any QSPI object, no matter where constructed. Using a pointer seems to make injection easier after we also add a new constructor, but this is fine for now.

rwalton-arm replied:
If we add a new constructor that allows us to inject a `QSPI` object, we should probably take a `unique_ptr<QSPI>`. Passing a `unique_ptr` semantically means that we're taking ownership of the `QSPI` object and managing the memory in this class. 

If we just take a raw pointer in the constructor we're changing this class to be non-owning with respect to `QSPI`, which could be a relatively big change. We'd also have an "owning" and a "non-owning" constructor which could make the internal implementation more complex.

LDong-Arm replied:
I was mainly worried if the QSPI object was a pointer, the heap would be used if the caller does not inject a preexisting QSPI instance. This would introduce heap usage while it wasn't needed before.

Patater replied:
Are QSPI or BlockDevice objects themselves usually made on the heap anyway? Heap pressure should be about the same. I see your point about not wanting to add a dependency on the heap where we didn't have one before. Does QSPIFBlockDevice not use the heap in other places?

LDong-Arm replied:
The only direct use of the heap by QSPIFBlockDevice is the CSEL array. And the SFDP sector map parsing function temporarily stores SFDP data in the heap too, and it's freed when parsing finishes (the unique_ptr goes out of scope).

rwalton-arm commented on storage/blockdevice/tests/UNITTESTS/SFDP/test_sfdp.cpp:
```diff
@@ -254,6 +234,26 @@ TEST_F(TestSFDP, TestSFDPReadFailure)
+     * Based on Cypress S25FS512S, modified to have one descriptor,
+     * twelve regions for test purpose.
+     */
+    const uint8_t sector_map_single_descriptor_twelve_regions[] {
```
Can this be `constexpr`?

LDong-Arm replied:
I initially thought `constexpr` was for compile-time evaluation, and couldn't think of how this particular test case would benefit from it in practice, as it's just passed around like an array/pointer?

rwalton-arm replied:
Plausibly it won't benefit us in this case. Yes, `constexpr` means the array is treated as a constant expression which is evaluated at compile-time, which is slightly different than `const` (which means it's read-only). If we were passing it to other objects that could possibly be `constexpr` we would see some performance gains, I think. This is because the operations on the array could happen at compile time, and the compiler would just do a direct substitution of the result instead of generating the lookup/usage code. Given the way our current code is written I don't think we will see those benefits; we use pointer indirection everywhere so we'll pay the cost of de-referencing the pointers at run time, also almost none of the calling code is `constexpr` so we'd probably need to rewrite it to be `constexpr` where possible to see any benefit.

LDong-Arm replied:
I'm leaving it as `const` for now, because this array emulates raw bytes that are normally fetched from a device at run time.

rwalton-arm commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -554,7 +554,7 @@ int QSPIFBlockDevice::get_erase_value() const
 static PinName *generate_initialized_active_qspif_csel_arr()
 {
-    PinName *init_arr = new PinName[QSPIF_MAX_ACTIVE_FLASH_DEVICES];
+    static PinName init_arr[QSPIF_MAX_ACTIVE_FLASH_DEVICES];
```
Is there a reason we can't use a `std::array` instead of a c-style array? Maybe we don't need value-semantics here as this is `static` anyway?

LDong-Arm replied:
Could you explain mean by value-semantics here?

As for `std::array`, I guess the benefits are like iterators and bound checks, etc. But the bound check can potentially throw `std::out_of_range` which we can't catch due to `-fno-exceptions`.

rwalton-arm replied:
By value-semantics I mean that the array itself will be treated as a value rather than decaying to a pointer to its first element. So, when passing it as an argument to a function we can pass the array by `const &` or by value instead of using pointer semantics. We don't have to pass a "size" argument with it. We don't have to keep track of the size at all. The size is actually part of the type. So, you could write a function that only takes an array of size N, which would fail to compile if we passed it an incorrectly sized array. It's generally good style to default to `std::array` if we have a fixed size array, rather than using the more error prone "C-style" array for those reasons. 

This was a question rather than a request for a change, anyway. As I mentioned, because this is static, maybe we don't need to copy or pass it as a parameter anywhere. It's good to at least consider using the more modern c++ constructs in as many cases as possible (they're generally stronger typed and safer/easier to use than the C equivalents), and I wasn't sure if we'd considered it.

LDong-Arm replied:
Totally agree. The limitation of the current implementation is, this function returns a pointer, which gets set as the value of `QSPIFBlockDevice::_active_qspif_flash_csel_arr` (a static pointer) so it gets back to C-style code again. But we're free to rework the internals of a class as long as it doesn't break any public API. I'd prefer making `QSPIFBlockDevice::_active_qspif_flash_csel_arr` a static array directly, so we won't lose size info. Will make the change.

LDong-Arm replied:
This would involve a bit of refactoring. I'll let other reviewers say if we want to do it in this PR.

Patater replied:
How much memory is this? I'd generally prefer to use the heap so memory can be used for other purposes if needed. Static hogs the memory, whether it's in use or not.

This is a static function, so we can't just free the memory in the QSPIFBlockDevice destructor, but it'd be good to understand how much memory is being used here (at least on one or two typical devices) and use that to help justify making this change. Please update the commit message "QSPIFBlockDevice: Use static memory for CSEL array" with this information.

LDong-Arm replied:
> How much memory is this? I'd generally prefer to use the heap so memory can be used for other purposes if needed. Static hogs the memory, whether it's in use or not.

I agree with this, so I've just removed this commit. Valgrind says that this buffer, once allocated from the heap, never gets freed. But I think it's quite common for an embedded system to be assumed to run indefinitely without any shut-down clean-up. So shall we ignore the report from Valgrind?

If we want to be extra rigorous and free the array properly, I think we should add an "instance counter" for QSPI flash - deallocate when instances go to 0, (re)allocate when the first instance is created. Not sure if this complexity is beneficial for an ever-running system.

(If we were to keep this commit, the static memory taken would be `10 * sizeof(enum)` which is 320 bytes or less depending on if the compiler optimises the enum, as the maximum supported enum is int.)

rwalton-arm commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -1099,6 +1102,15 @@ int QSPIFBlockDevice::_handle_vendor_quirks()
             _num_status_registers = 1;
             break;
+        case 0x01:
+            if (vendor_device_ids[1] == 0x02 && vendor_device_ids[2] == 0x20) {
```
All the other cases emit a trace debug message to say we're applying the "quirk". Should we also add this to our new case?

LDong-Arm replied:
Good idea

LDong-Arm replied:
Done

rwalton-arm commented on storage/blockdevice/source/SFDP.cpp:
```diff
@@ -291,13 +291,96 @@ int sfdp_parse_sector_map_table(Callback<int(bd_addr_t, sfdp_cmd_addr_size_t, ui
+    // Otherwise the configuration ID is 0x00.
+    uint8_t active_config_id = 0x00;
+    // If any configuration detection command descriptors exist, the table starts with one.
+    if (!(descriptor[0] & 0x02)) {
```
Should `0x01` and `0x02` be named variables? We use them more than once in this code.

LDong-Arm replied:
I'll name those if they are used multiple times. I tried not to alias any bit masks that are used once only, as the code should be self-explanatory (i.e. what they are for).

rwalton-arm commented on storage/blockdevice/source/SFDP.cpp:
```diff
@@ -291,13 +291,96 @@ int sfdp_parse_sector_map_table(Callback<int(bd_addr_t, sfdp_cmd_addr_size_t, ui
+            auto addr_size = static_cast<sfdp_cmd_addr_size_t>(descriptor[2] >> 6);
+            uint8_t mask = descriptor[3];
+            uint32_t cmd_addr;
+            memcpy(&cmd_addr, &descriptor[4], sizeof(cmd_addr)); // bytes 4-7 of the descriptor
```
Is this `memcpy` a performance optimisation? Can we use copy-assignment or `std::copy` to do this in a less error-prone way?

LDong-Arm replied:
The left hand side `cmd_addr` is a 32-bit unsigned integer, but the right hand side `descriptor` is an array of 8-bit unsigned integers. This kind of assignment is semantically only doable with `memcpy`, and it's the safest when alignment is not guaranteed.

rwalton-arm commented on storage/blockdevice/source/SFDP.cpp:
```diff
@@ -291,13 +291,96 @@ int sfdp_parse_sector_map_table(Callback<int(bd_addr_t, sfdp_cmd_addr_size_t, ui
+        }
+    }
+
+    // Part 2: sector map descriptors.
```
Does it make sense to factor out "part one" and "part 2" into separate helper functions with clear names?

LDong-Arm replied:
Good point. I wanted reviewers' opinion on breaking down this big function, now you confirmed it so I'll do it.

LDong-Arm replied:
Done

Patater replied:
In commit "SFDP: Add support for multiple configurations and sector maps", I still see "Part 1" and "Part 2" in the same function.

LDong-Arm replied:
The code was already refactored, but I kept the comments. Now I've rewritten the comments to not say "Part 1" and "Part 2".

rwalton-arm commented on storage/blockdevice/tests/TESTS/blockdevice/general_block_device/main.cpp:
```diff
@@ -563,69 +563,75 @@ void test_erase_functionality()
+
+        // Allocate buffer for write test data
+        size_t data_buf_size = erase_size;
+        uint8_t *data_buf = new (std::nothrow) uint8_t[data_buf_size];
```
Do we have `std::unique_ptr` available?

LDong-Arm replied:
Mbed OS doesn't seem to have written guidelines around the STL vs manual/C-style code, but it seems in practice uses of STL are relatively rare (but exists in a few places). Performance/overhead is one reason, but anything that may raise exceptions can't be used because of `-fno-exceptions`. If we do `make_unique()` to do allocation here, it can potentially throw `std::bad_alloc` which we can't catch and falls back to a crash. This prevents us from using a very large proportion of the STL.

It's a good question though. The existing code in SFDP.cpp uses `make_unique` and checks for `nullptr`, but this won't work so I should fix that.

rwalton-arm replied:
If we encounter an out of memory condition can we do anything else other than crash? I would've thought even trying to catch `std::bad_alloc` is extremely rare, as we can't really continue if we get a bad alloc anyway. Mbed OS in general seems to have a rather inconsistent usage of c++ vs c style code. I'd generally prefer to default to c++ when possible. 

I do see your point about the fact we have exceptions disabled so we lose any exception safety from `make_unique`. Could we not still use the RAII features of `unique_ptr` to guarantee this memory is freed instead of freeing it manually? We don't need to use the `make_unique` factory to create a `unique_ptr`, we can alloc the memory ourselves and the `unique_ptr` will ensure it's deleted. I don't see why `auto data_buf = std::unique_ptr<uint8_t[]>(new (std::nothrow) int[data_buf_size]);` wouldn't work, and then we still get to benefit from the automatic free.

LDong-Arm replied:
> If we encounter an out of memory condition can we do anything else other than crash?

Some dynamic allocations are not critical, so we should be able to catch them and return errors. Exceptions that are purely due to bugs in code (e.g. array out of bound) can be allowed to crash in my opinion.

> I don't see why auto data_buf = std::unique_ptr<uint8_t[]>(new (std::nothrow) int[data_buf_size]); wouldn't work, and then we still get to benefit from the automatic free.

Cool, I'll try that.

LDong-Arm replied:
> auto data_buf = std::unique_ptr<uint8_t[]>(new (std::nothrow) uint8_t[data_buf_size]);

This works! But this particular file does lots of dynamic allocations, so fixing them all would be refactoring, and changing this line only would make style inconsistent. I think we should create an issue for improving C++ coding in Mbed OS.

I am using this to fix the existing `std::unique_ptr` allocation check in SFDP.cpp though. Good suggestion!

LDong-Arm commented on storage/blockdevice/source/SFDP.cpp:
```diff
@@ -254,7 +268,7 @@ int sfdp_parse_sector_map_table(Callback<int(bd_addr_t, void *, bd_size_t)> sfdp
+     * are variable -> the size of this table is variable
      */
     auto smptbl_buff = std::make_unique<uint8_t[]>(sfdp_info.smptbl.size);
     if (!smptbl_buff) {
```
This doesn't work, because `make_unique` throws `std::bad_alloc` instead of returning, but we use `-fno-exceptions`.

LDong-Arm committed:
QSPIFBlockDevice: Add quirk for inconsistent CR3NV register value

Cypress S25FS512S's SFDP table suggests that bit-1 of the register
CR3NV on 25FS512S should equal 1. But it is a known issue that the
value is actually 0 on hardware. So if we query the value of CR3NV
during configuration detection, we can't find a configuration that
matches the SFDP data. This issue has been discussed in the Linux MTD
(Memory Technology Devices) mailing list:
https://linux-mtd.infradead.narkive.com/ylHK6CyT/spi-nor-fs512s-incorrect-cr3nv-1-value

This commit adds a quirk to report bit-1 of CR3NV as 1.

Note: In Mbed OS, vendor-specific quirks can only be handled in block
devices. The SFDP functions assume data from hardware to be correct.
So this quirk is in the block device.

LDong-Arm committed:
SFDP: Fix sector map table allocation check

When passing an allocation size directly to `std::make_unique`,
`std::nothrow` is unavailable, so any failed allocation results in
an exception which we cannot catch because Mbed OS is compiled with
`-fno-exceptions`. To fix this, chain `std::make_unique` with
`new (std::nothrow)`, and the allocated `unique_ptr` retains the
`nullptr` property.

Patater commented on storage/blockdevice/COMPONENT_SPIF/include/SPIF/SPIFBlockDevice.h:
```diff
@@ -221,10 +221,12 @@ class SPIFBlockDevice : public mbed::BlockDevice {
     /****************************************/
     // Send SFDP Read command to Driver
-    int _spi_send_read_sfdp_command(mbed::bd_addr_t addr, void *rx_buffer, mbed::bd_size_t rx_length);
+    int _spi_send_read_sfdp_command(mbed::bd_addr_t addr, mbed::sfdp_cmd_addr_size_t addr_size,
```
Shouldn't this change be done in the commit "SFDP: Add more parameters to the reader callback"? Otherwise, I don't think "SFDP: Add more parameters to the reader callback" will even build.

LDong-Arm replied:
Squashed them

Patater commented on storage/blockdevice/COMPONENT_OSPIF/include/OSPIF/OSPIFBlockDevice.h:
```diff
@@ -327,7 +327,9 @@ class OSPIFBlockDevice : public mbed::BlockDevice {
 
     // Send command to read from the SFDP table
-    int _ospi_send_read_sfdp_command(mbed::bd_addr_t addr, void *rx_buffer, mbed::bd_size_t rx_length);
+    int _ospi_send_read_sfdp_command(mbed::bd_addr_t addr, mbed::sfdp_cmd_addr_size_t addr_size,
```
Same goes for these other callbacks (should be part of "SFDP: Add more parameters to the reader callback")

LDong-Arm replied:
Done

Patater commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -17,15 +17,15 @@
 #include "blockdevice/internal/SFDP.h"
 #include "platform/Callback.h"
-#include "QSPIFBlockDevice.h"
+#include "QSPIF/QSPIFBlockDevice.h"
```
"QSPIFBlockDevice: Use fully-qualified include path" doesn't have a "why" explanation in the commit message. We should explain why this is good, or why we need to do it.

LDong-Arm replied:
Added

Patater commented on storage/blockdevice/COMPONENT_QSPIF/include/QSPIF/QSPIFBlockDevice.h:
```diff
@@ -376,6 +376,9 @@ class QSPIFBlockDevice : public mbed::BlockDevice {
 
     bool _needs_fast_mode;
 
+    // S25FS512S needs a quirk
```
In commit "QSPIFBlockDevice: Add quirk for inconsistent CR3NV register value", we say "But it is a known issue". Can you link to errata or mailing list posts for this in the commit message?

LDong-Arm replied:
Done, added a link to the mailing list. There is an erratum in the datasheet on this, but only says CR3NV[1] is "unused", but it is actually expected by SFDP, so the erratum isn't quite informative.

Patater commented on storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp:
```diff
@@ -1109,6 +1109,13 @@ int QSPIFBlockDevice::_handle_vendor_quirks()
                 // * The SFDP table expects the register bitfield CR3NV[1] to be 1
                 //   but its actual value on the hardware is 0. In order for SFDP parsing
                 //   to work, the quirk reports CR3NV[1] as 1.
+                // * All three possible configurations support 256KB sectors across
```
In commit "QSPIFBlockDevice: Add quirk to ignore overlaying sectors on S25FS512S", we say "if we changed them, they would become irreversible." -- could we even change them? If they are OTP, how could we change them? If CR1NV and CR3NV are not programmed, what value would we see?

LDong-Arm replied:
The factory values are 0, and each bit field can be flipped from 0 -> 1 but not 1 -> 0. So the registers are programmable but not erasable. I've updated the commit message.

(I think the idea is, the chip's manufacturer expects the user to do a permanent one-time configuration and stick with it. There's also a register to permanently lock the registers from all further changes (even 0 -> 1), but it's out of the scope and not added to the commit.)

Patater commented on storage/blockdevice/COMPONENT_QSPIF/UNITTESTS/test_QSPIFBlockDevice.cpp:
```diff
@@ -0,0 +1,285 @@
+    // sectors), as Mbed OS does not support this type of flash layout.
+    EXPECT_EQ(0x0, CR1NV & 0x4);
+    EXPECT_EQ(0x8, CR3NV & 0x8);
+}
```
Do we have a test to check that the quirk is not applied when the chip is not S25FS512S?

LDong-Arm replied:
Good point, just added

Patater commented on targets/targets.json:
```diff
@@ -7750,12 +7750,8 @@
             "UDB_SDIO_P12"
         ],
-        "components_remove": [
-            "QSPIF"
```
Yay!

jeromecoutant reviewed (approved):
Tests re-executed OK with STM32 targets

Patater reviewed (approved):
LGTM

rwalton-arm reviewed (approved):
LGTM
